File: site/testsuites/testsuite-kotlin/src/test/kotlin/docs/tbdex/issuer/CredentialIssuanceTest.kt
package website.tbd.developer.site.docs.tbdex

import com.nimbusds.jwt.JWTClaimsSet
import org.junit.jupiter.api.Assertions.*
import com.nimbusds.jwt.JWTParser
import io.ktor.server.routing.*
import io.ktor.server.application.*
import io.ktor.server.response.*
import io.ktor.server.request.*
import io.ktor.http.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import web5.sdk.dids.did.BearerDid
import web5.sdk.credentials.VerifiableCredential
import web5.sdk.crypto.InMemoryKeyManager
import web5.sdk.dids.methods.dht.CreateDidDhtOptions
import web5.sdk.dids.methods.dht.DidDht
import java.security.SignatureException
import io.ktor.server.routing.*
import io.ktor.server.application.*
import io.ktor.server.response.*
import io.ktor.server.request.*
import io.ktor.http.*
import kotlinx.coroutines.runBlocking

import web5.sdk.dids.didcore.Service



// :snippet-start: sanctionsCredentialClassKt
data class SanctionsCredential(val listsCleared: List<String>)
// :snippet-end:
// :snippet-start: checkSanctionsListsFunctionAndSanctionsListResultClassKt
fun checkSanctionsList(): SanctionsListResult {
    // Implement sanctions list check logic
    return SanctionsListResult(
        isSanctioned = false,
        listsCleared = listOf(
            "FBI's Most Wanted", "USA Watchlist", "EU Watchlist"
            )
        )
}

// sanctionsListResult class
data class SanctionsListResult(val isSanctioned: Boolean, val listsCleared: List<String>)
// :snippet-end:

private fun createADid() = runBlocking {
    // :snippet-start: createADidWithAServiceEndpointKT
    val idvService = Service(
        id = "idv",
        type = "IDV",
        serviceEndpoint = listOf("https://exampleIdvEndpoint.com/idv/siopv2/initiate")
    )

    val options = CreateDidDhtOptions(
        services = listOf(idvService),
        publish = true
    )

    val issuerBearerDid = DidDht.create(InMemoryKeyManager(), options)
    // :snippet-end:
}


class CredentialIssuanceTest {

    @Test
    fun `create() creates a credential with expected fields`() {
        val issuerDid = generateTestDid()
        val subjectDid = "did:dht:aq5ek9jbf8kota7ufcbgnnc4ikkfpd8b1u9on1b1n7k7wdcapbgo"

        val sanctionsListResult = SanctionsListResult(
            isSanctioned = false,
            listsCleared = listOf("FBI's Most Wanted", "USA Watchlist", "EU Watchlist")
        )

        val createdCredential = VerifiableCredential.create(
            type = "SanctionsCredential",
            issuer = issuerDid.uri.toString(),
            subject = subjectDid,
            data = sanctionsListResult
        )

        assertNotNull(createdCredential, "The created credential should not be null.")
        assertEquals("SanctionsCredential", createdCredential.type, "The credential type should match.")
        assertEquals(issuerDid.uri.toString(), createdCredential.issuer, "The issuer should match.")
    }

    @Test
    fun `sign() method signs the credential and returns a non-empty string`() {
        val issuerDid = generateTestDid()
        val subjectDid = "did:dht:aq5ek9jbf8kota7ufcbgnnc4ikkfpd8b1u9on1b1n7k7wdcapbgo"

        val sanctionsListResult = SanctionsListResult(
            isSanctioned = false,
            listsCleared = listOf("FBI's Most Wanted, USA Watchlist", "EU Watchlist")
        )

        val sanctionsCredential = VerifiableCredential.create(
            type = "SanctionsCredential",
            issuer = issuerDid.uri.toString(),
            subject = subjectDid,
            data = sanctionsListResult
        )

        val signedCredential = sanctionsCredential.sign(issuerDid)

        assertTrue(signedCredential.isNotEmpty(), "The signed credential should not be an empty string.")
    }

    @Test
    fun `verify() does not throw an exception for a valid JWT`() {
        val validJwt = "eyJ0eXAiOiJKV1QiLCJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3djSmkzeVVONDJFZ1N2cmNHRlFyQzRKY1pkenlWWkhQOVdmMXFRZWRuVlRQI3o2TWt3Y0ppM3lVTjQyRWdTdnJjR0ZRckM0SmNaZHp5VlpIUDlXZjFxUWVkblZUUCJ9.eyJ2YyI6eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSWRlbnRpZmljYXRpb25DcmVkZW50aWFsIl0sImlkIjoidXJuOnV1aWQ6MDQ3ZTg0ZWItY2NhMS00NjFlLWFjZjAtMGMyZGE5ZDczOTNhIiwiaXNzdWVyIjoiZGlkOmtleTp6Nk1rd2NKaTN5VU40MkVnU3ZyY0dGUXJDNEpjWmR6eVZaSFA5V2YxcVFlZG5WVFAiLCJpc3N1YW5jZURhdGUiOiIyMDI0LTAyLTE1VDE5OjMyOjE2WiIsImNyZWRlbnRpYWxTdWJqZWN0Ijp7ImlkIjoiZGlkOmtleTp6Nk1rd2NKaTN5VU40MkVnU3ZyY0dGUXJDNEpjWmR6eVZaSFA5V2YxcVFlZG5WVFAiLCJuYW1lIjoiam9obiJ9fSwiaXNzIjoiZGlkOmtleTp6Nk1rd2NKaTN5VU40MkVnU3ZyY0dGUXJDNEpjWmR6eVZaSFA5V2YxcVFlZG5WVFAiLCJzdWIiOiJkaWQ6a2V5Ono2TWt3Y0ppM3lVTjQyRWdTdnJjR0ZRckM0SmNaZHp5VlpIUDlXZjFxUWVkblZUUCJ9.7oFAPckx-vxCbbzKSk9bU7eXlnjBFvMborl9woHCbcvWaLt0LvTRuDfvGDPC24V9D1K5OFpTnnBiN5jtIOmbBg"

        assertDoesNotThrow {
            VerifiableCredential.verify(validJwt)
        }
    }


    private fun generateTestDid(): BearerDid {
        val issuerDid = DidDht.create(InMemoryKeyManager(), CreateDidDhtOptions(publish = true))
        return issuerDid
    }

    private fun Application.configureRouting() {
        val issuerDid = generateTestDid()
        // :snippet-start: checkSanctionsEndpointKt
        routing {
            get("/check-sanctions") {
                try {

                    /***********************************************
                    * Accepts a JWT in the Authorization header
                    * and parse to get the signer DID
                    ************************************************/
                    val authHeader = call.request.header("Authorization")

                    if (authHeader == null || !authHeader.startsWith("Bearer ")) {
                        call.respond(
                            HttpStatusCode.Unauthorized,
                            "Authorization header required"
                        )
                        return@get
                    }

                    val compactJwt = authHeader.removePrefix("Bearer ")
                    var signerDid: String? = null

                    try {
                        VerifiableCredential.verify(compactJwt)

                        val claimsSet: JWTClaimsSet = JWTParser.parse(compactJwt).jwtClaimsSet
                        signerDid = claimsSet.subject
                    } catch (e: SignatureException) {
                        println("JWT Verification failed: ${e.message}")
                        call.respond(HttpStatusCode.Unauthorized, "Invalid token")
                        return@get
                    }

                    signerDid?.let { did ->

                        /***********************************************
                        * Perform the sanctions check and get the result
                        ************************************************/
                        val sanctionsListResult = checkSanctionsList()

                        if (sanctionsListResult.isSanctioned) {
                            call.respond(
                                HttpStatusCode.Forbidden,
                                "User is not eligible for a credential due to sanctions"
                            )
                            return@get
                        }

                        /***********************************************
                        * Create credential
                        ************************************************/
                        val sanctionsCredential = VerifiableCredential.create(
                            type = "SanctionsCredential",
                            issuer = issuerDid.uri.toString(),
                            subject = did,
                            data = SanctionsCredential(sanctionsListResult.listsCleared)
                        )

                        /***********************************************
                        * To secure the VC, you must sign it
                        ************************************************/
                        val credentialToken = sanctionsCredential.sign(issuerDid)

                        call.respond(
                            HttpStatusCode.OK,
                            "Signed Credential Token: $credentialToken"
                        )
                    } ?: run {
                        call.respond(
                            HttpStatusCode.BadRequest,
                            "Signer DID could not be determined from the JWT"
                        )
                    }

                } catch (e: Exception) {
                    // Generic error handling
                    call.respond(
                        HttpStatusCode.InternalServerError,
                        "An unexpected error occurred: ${e.message}"
                    )
                }
            }
        }
        //:snippet-end:
    }

}

File: site-new/testsuites/testsuite-javascript/__tests__/tbdex/issuer/credentialIssuance.test.js
import { describe, test, expect } from 'vitest';
import express from 'express';
import { Jwt, VerifiableCredential } from '@web5/credentials';
import { DidDht } from '@web5/dids';

const issuerDid = await DidDht.create({
  options: { publish: true },
});

const signerDid = await DidDht.create({
  options: { publish: true },
});

// :snippet-start: sanctionsCredentialClass
class SanctionsCredential {
  constructor(check) {
    this.listsCleared = check.listsCleared;
  }
}

export default SanctionsCredential;
// :snippet-end:

// :snippet-start: checkSanctionsListsFunction
function checkSanctionsList(payload) {
  // This is where you would add the functionality to perform the actual sanction checks
  return {
    isSanctioned: false,
    listsCleared: ["FBI's Most Wanted, USA Watchlist", 'EU Watchlist'],
  };
}
// :snippet-end:

// :snippet-start: createADidWithServiceEndpointJS
const issuerBearerDid = await DidDht.create({
  options: {
    publish: true,
    services: [
      {
        id: 'idv',
        type: 'IDV',
        serviceEndpoint: 'https://exampleIdvEndpoint.com/idv/siopv2/initiate',
      },
    ],
  },
});
// :snippet-end:

const app = express();
app.use(express.json());

// :snippet-start: checkSanctionsEndpoint
app.get('/check-sanctions', async (req, res) => {
  /***********************************************
   * Accepts a JWT in the Authorization header
   * and parse to get the signer DID
   ************************************************/
  try {
    const authHeader = req.headers['authorization'];

    if (!authHeader) {
      return res.status(401).json({
        errors: ['Authorization header required'],
      });
    }

    const tokenParts = authHeader.split('Bearer ');

    if (tokenParts.length !== 2) {
      return res.status(401).json({
        errors: ['Authorization header format is Bearer <token>'],
      });
    }

    const compactJwt = tokenParts[1];

    let subject, payload;

    try {
      const verificationResult = await Jwt.verify({ jwt: compactJwt });
      subject = verificationResult.payload.sub;
      payload = verificationResult.payload;
    } catch (error) {
      return res.status(401).json({
        errors: ['Invalid token'],
      });
    }

    /***********************************************
     * Perform the sanctions check and get the result
     ************************************************/
    const sanctionsListResult = checkSanctionsList(payload);

    if (!sanctionsListResult || sanctionsListResult.isSanctioned) {
      return res.status(403).json({
        errors: ['User is not eligible for a credential due to sanctions'],
      });
    }

    /***********************************************
     * Create credential
     ************************************************/
    const sanctions_credential = await VerifiableCredential.create({
      type: 'SanctionsCredential',
      issuer: issuerDid.uri,
      subject: subject,
      data: new SanctionsCredential(sanctionsListResult),
    });

    /***********************************************
     * To secure the VC, you must sign it
     ************************************************/
    const credential_token = await sanctions_credential.sign({
      did: issuerDid,
    });

    return res.status(200).json({ credential: credential_token });
  } catch (error) {
    // Generic error handling
    return res.status(500).json({
      errors: [`An unexpected error occurred: ${error.message}`],
    });
  }
});
// :snippet-end:

/* 
Have to add tests separately because the big snippet is
contained within the express app and the objects within
are outside of the function's scope to be tested 
*/

describe('Sanctions Credential Issuance', () => {
  const sanctionsListResult = {
    listsCleared: ["FBI's Most Wanted, USA Watchlist", 'EU Watchlist'],
  };

  test('Jwt.verify() works with a valid JWT', async () => {
    const validJwt =
      'eyJ0eXAiOiJKV1QiLCJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3djSmkzeVVONDJFZ1N2cmNHRlFyQzRKY1pkenlWWkhQOVdmMXFRZWRuVlRQI3o2TWt3Y0ppM3lVTjQyRWdTdnJjR0ZRckM0SmNaZHp5VlpIUDlXZjFxUWVkblZUUCJ9.eyJ2YyI6eyJAY29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSJdLCJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSWRlbnRpZmljYXRpb25DcmVkZW50aWFsIl0sImlkIjoidXJuOnV1aWQ6MDQ3ZTg0ZWItY2NhMS00NjFlLWFjZjAtMGMyZGE5ZDczOTNhIiwiaXNzdWVyIjoiZGlkOmtleTp6Nk1rd2NKaTN5VU40MkVnU3ZyY0dGUXJDNEpjWmR6eVZaSFA5V2YxcVFlZG5WVFAiLCJpc3N1YW5jZURhdGUiOiIyMDI0LTAyLTE1VDE5OjMyOjE2WiIsImNyZWRlbnRpYWxTdWJqZWN0Ijp7ImlkIjoiZGlkOmtleTp6Nk1rd2NKaTN5VU40MkVnU3ZyY0dGUXJDNEpjWmR6eVZaSFA5V2YxcVFlZG5WVFAiLCJuYW1lIjoiam9obiJ9fSwiaXNzIjoiZGlkOmtleTp6Nk1rd2NKaTN5VU40MkVnU3ZyY0dGUXJDNEpjWmR6eVZaSFA5V2YxcVFlZG5WVFAiLCJzdWIiOiJkaWQ6a2V5Ono2TWt3Y0ppM3lVTjQyRWdTdnJjR0ZRckM0SmNaZHp5VlpIUDlXZjFxUWVkblZUUCJ9.7oFAPckx-vxCbbzKSk9bU7eXlnjBFvMborl9woHCbcvWaLt0LvTRuDfvGDPC24V9D1K5OFpTnnBiN5jtIOmbBg';
    const verificationResult = await Jwt.verify({ jwt: validJwt });

    expect(verificationResult).toHaveProperty('header');
    expect(verificationResult).toHaveProperty('payload');
    expect(verificationResult.payload).toHaveProperty('iss');
    expect(verificationResult.payload).toHaveProperty('sub');
  });

  test('add service endpoint to existing DID document', async () => {
    // :snippet-start: updateADidWithServiceEndpointJS
    issuerBearerDid.document.service.push({
      id: 'idv',
      type: 'IDV',
      serviceEndpoint: 'https://exampleIdvEndpoint.com/idv/siopv2/initiate',
    });
    
    const updatedDidDocument = await DidDht.publish({ did: issuerBearerDid });
    // :snippet-end:

    expect(updatedDidDocument.didDocument).toHaveProperty('service');
    expect(updatedDidDocument.didDocument.service.some(service => service.id === 'idv')).toBeTruthy();
  });

  test('.create() creates a credential with expected fields', async () => {
    const createdCredential = await VerifiableCredential.create({
      type: 'SanctionsCredential',
      issuer: issuerDid.uri,
      subject: signerDid.uri,
      data: new SanctionsCredential(sanctionsListResult),
    });

    expect(createdCredential).toHaveProperty('type');
    expect(createdCredential).toHaveProperty('issuer', issuerDid.uri);
  });

  test('.sign() method signs the credential and includes expected fields', async () => {
    const sanctionsCredential = await VerifiableCredential.create({
      type: 'SanctionsCredential',
      issuer: issuerDid.uri,
      subject: signerDid.uri,
      data: new SanctionsCredential(sanctionsListResult),
    });

    const signedCredential = await sanctionsCredential.sign({
      did: issuerDid,
    });

    expect(typeof signedCredential).toBe('string');
    expect(signedCredential).not.toBe('');
  });
});
